src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -249,3 +473,226 @@ def calculate_ramp_position_and_duration(
+        exit_velocities[motor] = (
+            2 * (midpoint_upper_distance / half_durations[gap]) - exit_midpoint_velocity
+        )
+
+        # Travel distance across gap
+        distances[motor] = slice.lower[motor][gap] - slice.upper[motor][gap - 1]
+        if np.isclose(distances[motor], 0.0, atol=1e-12):
+            distances[motor] = 0.0
+
+    return entry_velocities, exit_velocities, distances
+
+
+def _calculate_profile_from_velocities(
+    motors: list[Motor],
+    slice: Slice,
+    gap: int,
+    time_arrays: dict[Motor, npt.NDArray[np.float64]],
+    velocity_arrays: dict[Motor, npt.NDArray[np.float64]],
+) -> GapSegment:
+    """Convert per-axis time/velocity profiles into aligned time/position profiles.
+
+    Given per-axis arrays of times and corresponding velocities,
+    this builds a single unified timeline containing all unique velocity change points
+    from every axis. It then steps through that timeline, and for each axis:
+
+    * If the current time matches one of the axis's own velocity change points,
+        use that known velocity.
+    * Otherwise, linearly interpolate between the axis's previous and next
+        known velocities, based on how far through that section we are.
+
+    At each unified time step, the velocity is integrated over the step duration
+    (using the trapezoidal rule) to update the axis's position. This produces
+    per-axis position and velocity arrays that are aligned to the same global
+    time grid.
+
+    Example:
+        combined_times = [0.1, 0.2, 0.3, 0.4]
+        axis_times[motor] = [0.1, 0.4]
+        velocity_array[motor] = [2, 5]
+
+        At 0.1 → known vel = 2 (use directly)
+        At 0.2 → 1/3 of the way to next vel → 3.0
+        At 0.3 → 2/3 of the way to next vel → 4.0
+        At 0.4 → known vel = 5 (use directly)
+
+        These instantaneous velocities are integrated over each Δt to yield
+        positions aligned with the global timeline.
+    """
+    # Combine all per-axis time points into a single sorted arrat of times
+    # This way we can evaluate each motor along the same timeline
+    # We know all axes positions at t=0, so we drop this point
+    combined_times = np.sort(np.unique(np.concatenate(list(time_arrays.values()))))[1:]
+
+    # We convert a list of t into a list of Δt
+    # We do this by substracting against previous cumulative time
+    time_intervals = np.diff(np.concatenate(([0.0], combined_times))).tolist()
+
+    # We also now all axes positions when t=t_final, so we drop this point
+    # However, we need the interval for the next collection window, so we store it
+    *time_intervals, final_interval = time_intervals
+    combined_times = combined_times[:-1]
+    num_intervals = len(time_intervals)
+
+    # Prepare dicts for the resulting position and velocity profiles over the turnaround
+    turnaround_profile: dict[Motor, npt.NDArray[np.float64]] = {}
+    turnaround_velocity: dict[Motor, npt.NDArray[np.float64]] = {}
```

Comment:
I would call these positions and velocities as that is what field you pass them to in `GapSegmment`